packages/compiler/src/install/install.ts

```diff
@@ -0,0 +1,198 @@
+  constructor(message: string) {
+    super({
+      code: "install-package-manager-error",
+      message,
+      severity: "error",
+      target: NoTarget,
+    });
+  }
+}
+
+export interface InstallTypeSpecDependenciesOptions {
+  readonly directory: string;
+  readonly stdio?: "inherit" | "pipe";
+  /** When set to true update the packageManager field with the package manger version and hash */
+  readonly savePackageManager?: boolean;
+}
+
+async function resolvePackageManagerSpecOrFail(
+  host: CliCompilerHost,
+  tracer: Tracer,
+  options: InstallTypeSpecDependenciesOptions,
+): Promise<[ResolvedSpecResult, readonly Diagnostic[]]> {
+  const result = await resolvePackageManagerSpec(host, tracer, options.directory);
+
+  switch (result.kind) {
+    case "no-package":
+      throw new InstallDependenciesError("No package.json found, cannot install dependencies.");
+    case "no-spec":
+      return [
+        {
+          kind: "resolved",
+          spec: {
+            name: "npm",
+            range: "latest",
+          },
+          path: result.path,
+        },
+        options.savePackageManager
+          ? [
+              {
+                code: "no-package-manager-spec",
+                severity: "warning",
+                message:
+                  "No package manager spec found, defaulted to npm latest version. Please set devEngines.packageManager or packageManager in your package.json.",
+                target: NoTarget,
+              },
+            ]
+          : [],
+      ];
+    case "resolved":
+      return [result, []];
+  }
+}
+
+export async function installTypeSpecDependencies(
+  host: CliCompilerHost,
+  options: InstallTypeSpecDependenciesOptions,
+): Promise<readonly Diagnostic[]> {
+  const { directory, stdio = "inherit", savePackageManager } = options;
+  const diagnostics = createDiagnosticCollector();
+  const tracer = createTracer(host.logger).sub("install");
+  try {
+    const { spec, path: packageJsonPath } = diagnostics.pipe(
+      await resolvePackageManagerSpecOrFail(host, tracer, options),
+    );
+    const packageManager = spec.name;
+    const packageManagerConfig = getPackageManagerConfig(packageManager);
+    const manifest = await fetchPackageManifest(packageManager, spec.range);
+    tracer.trace(
+      "fetched-manifest",
+      `Resolved manifest for ${packageManager} at version ${manifest.version}`,
+    );
+
+    const installDir = joinPaths(pmDir, packageManager, manifest.version);
+    await rm(installDir, { recursive: true, force: true });
+    await mkdir(installDir, { recursive: true });
+    const tempDir = await mkTempDir(host, pmDir, `tsp-pm-${packageManager}-${manifest.version}`);
+
+    tracer.trace(
+      "downloading-extracting",
+      `Downloading and extracting ${packageManager} at version ${manifest.version} in ${tempDir}`,
+    );
+    const extractResult = await downloadAndExtractPackage(manifest, tempDir, spec.hash?.algorithm);
```

Comment:
Looks like we need to check if the package manager is already installed before reacquiring here. This may involve some logic to see if the installed version satisfies the constraint or is up to date with the manifest.